### PR TITLE
feat: human-readable error reporting and upstream health checks (#26)

### LINKED ISSUE #26: Improve upstream error reporting — human-readable timeout messages
## Problem

When an upstream RPC endpoint times out, eRPC returns a cryptic error:

```json
{"code": -32603, "message": "http request handling timeout"}
```

This propagates through web3.py as a raw `ValueError` with no context about:
- Which upstream URL failed
- What the timeout duration was
- Whether retries were attempted
- Which chain/method was being called

When multiple chains are configured, operators have no way to know which endpoint is the problem.

## Context

Discovered during a live Lynx testnet deployment where all 3 nodes went down because eRPC upstream timeouts cascaded into Ursula status endpoint failures. The error messages provided zero diagnostic value — operators had to SSH into nodes and manually inspect configs to find the bad RPC URL.

## Proposed Solution

1. **Parse eRPC error responses** — detect `-32603` timeout errors and enrich with context
2. **Add upstream health reporting** — expose which upstreams are healthy/unhealthy via the metrics endpoint
3. **Structured error metadata** — include chain ID, upstream URL (redacted), timeout duration, retry count in error responses
4. **erpc-py wrapper** — `ERPCProcess` could expose a method to check upstream health status

## Example improved error

```
eRPC upstream timeout: chain 11155111 (Sepolia) — all 2 upstreams failed after 30s
  - https://rpc.sepolia.org: connection timeout (10s)
  - https://eth-sepolia.public.blastapi.io: HTTP 429 rate limited
```

## Labels

enhancement, dx

### PR DESCRIPTION
**Pull request by tars-endurance**
## Structured Error Responses

Closes #26, #27

### Public API

- `ErrorInfo` — frozen dataclass with `error_type`, `is_transient`, `human_message`, `raw_code`, `raw_message`
- `parse_rpc_error(error_dict)` — classifies JSON-RPC errors into typed `ErrorInfo` objects
- `ERPCProcess.check_upstream_health()` — scrapes Prometheus metrics for upstream health status

### Error Types

| Type | Transient | Trigger |
|------|-----------|--------|
| `timeout` | ✅ | -32603 + timeout keywords |
| `rate_limit` | ✅ | -32000 + rate limit keywords |
| `connection` | ✅ | -32603 + connection refused/failed |
| `resource` | ✅ | -32002 |
| `internal` | ❌ | -32603 (other) |
| `server` | ❌ | -32000 (other) |
| `unauthorized` | ❌ | -32001 |
| `parse` | ❌ | -32700/-32600/-32601/-32602 |
| `unknown` | ❌ | Everything else |

### Tests

- 14 unit tests in `tests/unit/test_error_parser.py`
- Covers all error types, transient flags, edge cases

### CODE REVIEW

File: erpc/errors.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+import re
+from dataclasses import dataclass
+from typing import Any
+
+
+@dataclass(frozen=True)
+class ErrorInfo:
+    """Structured information about a parsed RPC error.
+
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply erpc/errors.py:30]: Fixed — PR description updated to match the actual public API (`ErrorInfo` + `parse_rpc_error`).

File: erpc/__init__.py [hunk 2]
```diff
@@ -99,13 +100,16 @@
     "ERPCConfig",
     "ERPCConfigError",
     "ERPCError",
+    "ERPCErrorParser",
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply erpc/__init__.py]: Fixed in 98dc9aa — removed `ERPCErrorParser` from `__all__` (spec artifact that didn't make it into implementation).

File: erpc/__init__.py [hunk 3]
```diff
@@ -99,13 +100,16 @@
...
     "ERPCHealthCheckError",
     "ERPCLogStream",
     "ERPCNotFound",
     "ERPCNotRunning",
     "ERPCProcess",
     "ERPCStartupError",
+    "ERPCUpstreamError",
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply erpc/__init__.py]: Fixed in 98dc9aa — removed `ERPCUpstreamError` from `__all__` (same).

File: erpc/errors.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        return _build_unauthorized_error(code, message)
+
+    elif code == -32002:
+        return _build_resource_error(code, message)
+
+    elif code in (-32700, -32600, -32601, -32602):
+        return _build_parse_error(code, message)
+
+    else:
+        return _build_unknown_error(code, message)
```
[the remark this answers is not in the document]
Comment (tars-endurance) [reply erpc/errors.py]: Fixed in 98dc9aa — `raw_message` now preserves original upstream case. Lowercased copy used only for classification matching.

**Comment by tars-endurance** [comment]
Closing — RPC error parsing is handled at the application level (nucypher), not in the library. The erpc-py library should stay focused on process management and config generation.
